File: 프로그래머스/lv2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.js
function solution(orders, course) {
    var answer = [];
    let menu = new Map();

    function combination(order, idx, len, prev) {
        if (prev.length === len) {
            let cur_key = prev.sort().join('');
            if (menu.has(cur_key)) {
                menu.set(cur_key, menu.get(cur_key) + 1);
            } else menu.set(cur_key, 1);
            return;
        }
        
        for (let i = idx; i < order.length; i++) {
            combination(order, i + 1, len, [...prev, order[i]]);
        }
    }

    orders.map(order => {
        course.map(num => combination(order, 0, num, []));
    })

    menu = [...menu.entries()].filter(item => item[1] > 1).sort((a, b) => b[1] - a[1]).sort((a, b) => a[0].length - b[0].length);

    for (const num of course) {
        let max = 0
        let i = 0;
        for (const [x, y] of menu) {
            if (x.length != num) {
                menu.splice(0, i)
                break;
            }
            if (max == 0) {
                max = y;
            }
            if (y == max) {
                answer.push(x)
            }
            i += 1;
        }
    }

    return answer.sort();
}

File: 프로그래머스/lv2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
from collections import Counter
def solution(orders, course):
    answer = []

    for i in course:
        ans = []
        for j in orders:
            for x in combinations(j, i):
                x = list(x)
                x.sort(key=lambda x:x[0])
                ans.append("".join(x))
        max = Counter(ans).most_common()
        answer += [x for x, y in max if y > 1 and y == max[0][1]]

    return sorted(answer)